Add Milvus database compatibility with the RAG recipe

Shreyanand opened the pull request:
This PR adds the following changes:

- Adds a shell script to deploy a standalone Milvus container on the local system
- Seperate the vector db management logic from rag_chat.py file
- Add a logic to switch between Milvus or Chromadb depending on what is being used

@MichaelClifford I need to add documentation around this but I have tested it on my system and it works. What should be the next things to check to make sure this is polished?

rhatdan commented:
@Shreyanand you need to sign your commits.

git commit -a --amend -s
git push --force

rhatdan commented:
@MichaelClifford PTAL

MichaelClifford commented on recipes/natural_language_processing/rag/app/manage_vectordb.py:
```diff
@@ -0,0 +1,98 @@
+    #         hc = collection.count() < 1
+    #     elif self.vector_vendor == "milvus":
+    #         hc =
+    #     return hc
```
please remove this dead code

MichaelClifford commented on recipes/natural_language_processing/rag/app/manage_vectordb.py:
```diff
@@ -0,0 +1,98 @@
+
+    def create_retriever(self):
+        # Implement logic to create and return a retriever object from the VectorDB
+        print("Creating retriever object from VectorDB...")
```
Remove this, doesn't appear to do anything.

MichaelClifford commented on recipes/natural_language_processing/rag/app/manage_vectordb.py:
```diff
@@ -0,0 +1,98 @@
+    
+    def clear_db(self):
+        # Implement logic to clear the entire VectorDB
+        print(f"Clearing the entire VectorDB...")
```
```suggestion
        print(f"Clearing VectorDB...")
```

MichaelClifford commented on recipes/natural_language_processing/rag/app/manage_vectordb.py:
```diff
@@ -0,0 +1,98 @@
+        return db
+    
+    def clear_db(self):
+        # Implement logic to clear the entire VectorDB
```
```suggestion

```

MichaelClifford commented on recipes/natural_language_processing/rag/app/manage_vectordb.py:
```diff
@@ -0,0 +1,98 @@
+                print("DB populated")
+            else:
+                print("DB already populated")
+            db = Chroma(client=self.client,
```
Do we want to implement the same `.from_documents()` here like below with milvus? https://python.langchain.com/docs/integrations/vectorstores/chroma/

MichaelClifford commented on recipes/natural_language_processing/rag/app/rag_app.py:
```diff
@@ -1,74 +1,61 @@
+# vdb_vendor = os.getenv("VECTORDB_VENDOR", "milvus")
+# vdb_host = os.getenv("VECTORDB_HOST", "127.0.0.1")
+# vdb_port = os.getenv("VECTORDB_PORT", "19530")
+# vdb_name = os.getenv("VECTORDB_NAME", "test_collection")
```
```suggestion

```

MichaelClifford commented on recipes/natural_language_processing/rag/app/rag_app.py:
```diff
@@ -1,74 +1,61 @@
+def create_tmp_file():
+    with open(data.name, mode='wb') as w:
+        w.write(data.getvalue())
+    print(f"{data.name} uploaded")
```
```suggestion

```
remove I don't see this being used anywhere.

MichaelClifford commented on recipes/natural_language_processing/rag/app/rag_app.py:
```diff
@@ -1,74 +1,61 @@
+    with open(data.name, mode='wb') as w:
+        w.write(data.getvalue())
+    print(f"{data.name} uploaded")
+    documents = get_docs(data.name)
```
I think this will break on a permissions error if you try to write this file to disk in a container. See how we handled it with `tempfile` in this PR. https://github.com/containers/ai-lab-recipes/pull/330/files

MichaelClifford commented on recipes/natural_language_processing/rag/app/rag_app.py:
```diff
@@ -1,74 +1,61 @@
+    documents = get_docs(data.name)
 else:
-    print("DB already populated")
+    documents = get_docs("../sample-data/fake_meeting.txt")
```
```suggestion

```
I don't think we want to reset `documents` to this hard coded value when `data == None`. I'd just print that the vdb is empty.

Shreyanand replied:
If `data == None` then what will the `retriever` object be? The app UI shows an error in the llm chain when it is none. However, once you upload the doc the error goes from the app. Do we want to keep that workflow? Or am I doing something wrong?

MichaelClifford replied:
If there is no data, then the DB would be empty. The retriever would still get instantiated, just to an empty DB.  

I've not seen the error you are referencing.

In any case, I don't think defaulting to a hardcoded file name is the right solution.

Shreyanand replied:
I looked into instantiating the retriever to an empty DB, will update it. 
Agreed on the hardcoded file name, that is not a viable solution.

MichaelClifford commented on vector_dbs/milvus/milvus_standalone_embed.sh:
Can we convert this to a makefile or add it to docs instead of using this sh file? Do we actually need anything besides the podman run command here?

MichaelClifford commented:
@Shreyanand You will also need to update `recipes/natural_language_processing/rag/app/Containerfile` to include your new `manage_vectordb.py` file

Gregory-Pereira commented on vector_dbs/milvus/Makefile:
```diff
@@ -0,0 +1,55 @@
+
+.PHONY: run
+run:
+	podman run -d \
```
@Shreyanand I pulled this from your example, but the upstream script runs with `sudo`. I cannot find any reference to why it was put there / if it is actually required. Can you confirm that it ran worked as a standard user?

Gregory-Pereira replied:
Also @Shreyanand please sign your commits. I set DCO to pass just to see if that was blocking RAG tests from running but no dice there. Until then /hold

Shreyanand replied:
@Gregory-Pereira I tried it without sudo and it worked on my local. I am updating the PR with the reviews and DCO checks.

Gregory-Pereira reviewed (commented):
Did not have time to actually run the rag and make sure it works, I just focused on the DB parts and setting up the `Makefile`. I defer to Michael for the actual Rag and DB population bits.

rhatdan commented:
Needs a rebase.

rhatdan commented:
Also needs DCO
git commit -a --amend -s
git push --force

Shreyanand committed:
Add milvus vector database for rag recipe

Signed-off-by: Shreyanand <shanand@redhat.com>
Co-authored-by: Michael Clifford <mcliffor@redhat.com>
Co-authored-by: greg pereira <grpereir@redhat.com>

Gregory-Pereira commented on .github/workflows/rag.yaml:
```diff
@@ -5,16 +5,16 @@ on:
-      - ./recipes/natural_language_processing/rag/**
-      - .github/workflows/rag.yaml
+      - 'recipes/common/Makefile.common'
+      - '/recipes/natural_language_processing/rag/**'
```
```suggestion
      - 'recipes/natural_language_processing/rag/**'
```

Gregory-Pereira commented on recipes/natural_language_processing/rag/app/rag_app.py:
```diff
@@ -126,4 +98,4 @@ def read_file(file):
     st.chat_message("assistant").markdown(response.content)    
     st.session_state.messages.append({"role": "assistant", "content": response.content})
-    st.rerun()
+    st.rerun()
```
linting
```suggestion
    st.rerun()

```

Gregory-Pereira reviewed (commented):
/lgtm, but 2 tiny nits

Shreyanand committed:
Update .github/workflows/rag.yaml

Co-authored-by: Gregory Pereira <grpereir@redhat.com>

Update recipes/natural_language_processing/rag/app/rag_app.py

Co-authored-by: Gregory Pereira <grpereir@redhat.com>

Update Readme and add review comments

Signed-off-by: Shreyanand <shanand@redhat.com>

Gregory-Pereira reviewed (approved):
/lgtm

Gregory-Pereira commented:
Commit suggestions dont sign with DCO, major bummer. You will have to rebase those commit with signing. Personally I use: 
```
which gamens
gamens: aliased to git commit --amend --no-edit -S -s
```
Alternatively we can just set DCO to pass, but I would prefer stuff stay signed if possible.

MichaelClifford commented on vector_dbs/milvus/Makefile:
```diff
@@ -0,0 +1,55 @@
+
+.PHONY: run
+run:
+	podman run -d \
```
```suggestion
	podman run -it \
```
I'd prefer to see the container output by default 

MichaelClifford commented on recipes/natural_language_processing/rag/README.md:
```diff
@@ -4,7 +4,7 @@ This demo provides a simple recipe to help developers start to build out their o
 There are a few options today for local Model Serving, but this recipe will use [`llama-cpp-python`](https://github.com/abetlen/llama-cpp-python) and their OpenAI compatible Model Service. There is a Containerfile provided that can be used to build this Model Service within the repo, [`model_servers/llamacpp_python/base/Containerfile`](/model_servers/llamacpp_python/base/Containerfile).
 
-In order for the LLM to interact with our documents, we need them stored and available in such a manner that we can retrieve a small subset of them that are relevant to our query. To do this we employ a Vector Database alongside an embedding model. The embedding model converts our documents into numerical representations, vectors, such that similarity searches can be easily performed. The Vector  ...
+In order for the LLM to interact with our documents, we need them stored and available in such a manner that we can retrieve a small subset of them that are relevant to our query. To do this we employ a Vector Database alongside an embedding model. The embedding model converts our documents into numerical representations, vectors, such that similarity searches can be easily performed. The Vector  ...
```
@Shreyanand Since the recipe is defined in the ai-lab.yaml file and should only have One vectorDB. I think we should only reference Milvus here and move any mentions of chroma to a chroma readme. Also you should update the ai-lab.yaml file to reference this DB instead.

MichaelClifford commented on recipes/natural_language_processing/rag/app/rag_app.py:
```diff
@@ -1,91 +1,68 @@
 model_service = f"{model_service}/v1"
 chunk_size = os.getenv("CHUNK_SIZE", 150)
 embedding_model = os.getenv("EMBEDDING_MODEL","BAAI/bge-base-en-v1.5")
+vdb_vendor = os.getenv("VECTORDB_VENDOR", "chromadb")
```
lets make milvus the new default.

MichaelClifford commented on vector_dbs/README.md:
We should follow the same format as the `model_servers/ ` remove this doc, and add a README under each vectorDB directory with all the info to run it.

MichaelClifford commented on vector_dbs/milvus/Containerfile:
```diff
@@ -0,0 +1,2 @@
+FROM docker.io/milvusdb/milvus:master-20240426-bed6363f
```
```suggestion
FROM docker.io/milvusdb/milvus:master-20240516-5b27a0cd
```

MichaelClifford commented on recipes/natural_language_processing/rag/README.md:
You should also add to the readme the step needed to run the application with the new vectorDB.  I think the Env Variables are a little different now.
